sstables/storage.cc

```diff
@@ -36,10 +36,57 @@
 
 namespace sstables {
 
+class dir_wrapper final {
+    std::filesystem::path _pathname;
+    file _file;
+    bool _file_stale;
+
+public:
+    explicit dir_wrapper(std::filesystem::path pathname) : _pathname(std::move(pathname)) {};
+    dir_wrapper(const dir_wrapper&) = delete;
+    dir_wrapper& operator=(const dir_wrapper&) = delete;
+    dir_wrapper(dir_wrapper&&) = delete;
+    dir_wrapper& operator=(dir_wrapper&&) = delete;
+    ~dir_wrapper() = default;
+
+    const std::filesystem::path::string_type& native() const noexcept {
+        return _pathname.native();
+    }
+
+    operator std::filesystem::path() const noexcept {
+        return _pathname;
+    }
+
+    dir_wrapper& operator=(const sstring &nd) {
+        _pathname = nd;
+        _file_stale = true;
+        return *this;
+    }
+
+    future<> sync(const sstable& sst) {
```

Comment:
Seems to me this should be a low-level class, way below sstable, and not know anything about it.

Reply:
Right, I'll do that in v5. Instead of taking `sst` for the sake of `sst.sstable_write_io_check`, we can take `sst._write_error_handler`, and call `do_io_check`.

At the same time, I'm not really comfortable with all these sstable const refs being passed around between co-routines. There seems to be a some silent invariant that guarantees that the sst survives at least until the storage coroutines finish (or put differently, until the futures output by the storage member functions resolve). But if we lower this interface to take a more general error handler, I don't think it shoudl take a const ref to an error handler, but a copy. It's just a `std::function<void (std::exception_ptr)>`; it should not take up much space.

Reply:
Addressed in v5, documented in the v5 changelog.